File: paddle/api/Parameter.cpp
/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "paddle/parameter/Parameter.h"
#include "PaddleAPI.h"
#include "PaddleAPIPrivate.h"

Parameter::Parameter() : m(new ParameterPrivate()) {}

Parameter::~Parameter() { delete m; }

Parameter* Parameter::createFromRawPtr(void* ptr) {
  auto p = new Parameter();
  p->m->rawPtr = *static_cast<paddle::Parameter**>(ptr);
  return p;
}

Parameter* Parameter::createFromSharedPtr(void* ptr) {
  auto& p = *(paddle::ParameterPtr*)(ptr);
  if (p == nullptr) {
    return nullptr;
  } else {
    auto retParam = new Parameter();
    retParam->m->sharedPtr = p;
    return retParam;
  }
}

std::string Parameter::getName() const { return m->getPtr()->getName(); }

Vector* Parameter::getBuf(ParameterType type) {
  auto buf = m->getPtr()->getBuf(type);
  return Vector::createByPaddleVectorPtr(&buf);
}

ParameterConfig* Parameter::getConfig() {
  if (m->sharedPtr) {
    return ParameterConfig::createParameterConfigFromParameterSharedPtr(
        &m->sharedPtr);
  } else {
    return ParameterConfig::createParameterConfigFromParameterPtr(m->rawPtr);
  }
}

size_t Parameter::getID() const { return m->getPtr()->getID(); }

void Parameter::setValueUpdated() { m->getPtr()->setValueUpdated(); }

bool Parameter::save(const std::string& filename) const {
  return m->getPtr()->save(filename);
}

bool Parameter::load(const std::string& filename) const {
  return m->getPtr()->load(filename);
}

size_t Parameter::getSize() const { return m->getPtr()->getSize(); }

File: python/paddle/nn/parameter.py
# Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import TYPE_CHECKING

from paddle.base.framework import EagerParamBase
from paddle.tensor.creation import to_tensor

if TYPE_CHECKING:
    from paddle import Tensor


class Parameter(EagerParamBase):
    """
    Parameter is a subclass of Tensor, which is a persistable Tensor
    that can be updated by optimizers during training.

    Args:
        data (Tensor, optional): The initial data for the Parameter.
            If None, an empty Tensor will be created. Default: None.
        requires_grad (bool, optional): Whether this Parameter requires gradient computation.
            If True, the Parameter will accumulate gradients during backward pass.
            Default: True.

    Examples:
        .. code-block:: python

            >>> import paddle
            >>> # Create a Parameter from existing Tensor
            >>> weight = paddle.to_tensor([1.0, 2.0, 3.0])
            >>> param = paddle.nn.Parameter(weight)
            >>> print(param)

            >>> # Create a Parameter without initial data
            >>> param = paddle.nn.Parameter()
            >>> print(param)
    """

    def __init__(
        self, data: Tensor | None = None, requires_grad: bool = True
    ) -> Parameter:
        if data is None:
            data = to_tensor([])
        super().__init__(data.shape, data.dtype, trainable=requires_grad)
        super()._set_impl(data)
        self._is_param = True

    def __repr__(self) -> str:
        return super().__repr__()

    __str__ = __repr__
